README.md
# Optimizely Express Middleware

This repository houses the [Express.js](http://expressjs.com/) integration for use with Optimizely Full Stack and Optimizely Rollouts.

Optimizely Full Stack is A/B testing and feature flag management for product development teams. Experiment in any application. Make every feature on your roadmap an opportunity to learn. Learn more at https://www.optimizely.com/platform/full-stack/, or see the [documentation](https://docs.developers.optimizely.com/full-stack/docs).

Optimizely Rollouts is free feature flags for development teams. Easily roll out and roll back features in any application without code deploys. Mitigate risk for every feature on your roadmap. Learn more at https://www.optimizely.com/rollouts/, or see the [documentation](https://docs.developers.optimizely.com/rollouts/docs).

## Installation
Using NPM:
```
npm install @optimizely/express-sdk --save
```

Using yarn:
```
yarn add @optimizely/express-sdk
```

## Configuration
```javascript
const optimizelyExpressSdk = require('@optimizely/express-sdk');
const optimizely = optimizelyExpressSdk.initialize({
  sdkKey: 'CZsVVgn6j9ce6fNPt2ZEiB',
  datafileOptions: {
    autoUpdate: true,
    updateInterval: 600000 // 10 minutes in milliseconds
  },
});
```

The `autoUpdate` parameter indicates that your feature flags will get automatically updated from changes made in Optimizely's user interface. The `updateInterval` indicates how frequently the Optimizely Express SDK will poll for changes.

## Usage

### `optimizely.middleware`
```
app.use(optimizely.middleware);
```
The above middleware adds the following object on the express request object

**req.optimizely**:
```
req.optimizely = {
  datafile: A representation of all of your feature flags and experiments as defined in Optimizely
  client: The Optimizely SDK client instance which has methods like for `isFeatureEnabled`, `activate`, `track`, etc.
}
```

Once the middleware is installed, you can use the standard Optimizely JavaScript SDK APIs like [isFeatureEnabled](https://docs.developers.optimizely.com/rollouts/docs/is-feature-enabled)

```
app.get('/', function(req, res, next) {
  const isEnabled = req.optimizely.client.isFeatureEnabled('checkout_flow', '123');
  res.render('index', {
    title: 'Express: ' + (isEnabled ? 'feature on!' : 'feature off')
  });
});
```

### `optimizely.webhookRequest`
If you would rather that your server not poll for changes made in Optimizely's configuration but rather get updated as quickly as possible when changes occur, you can use the following secure webhook implementation. Note, this requires following the documentation on setting up a [secure webhook](https://docs.developers.optimizely.com/rollouts/docs/webhooks) in the Optimizely UI.
```
app.use('/webhooks/optimizely', bodyParser.text({ type: '*/*' }), optimizely.webhookRequest);
```
Once you have successfully implemented the webhook, you can turn off the polling completely by setting the `autoUpdate` attribute to false above in the [configuration](#Configuration) step.

Note: If you deploy your server to multiple different machines, this will not ensure that the two machines are in-sync with the latest configuration. If you would like to see support for cross-machine syncing via webhooks, please let us know by opening an issue on this repository.

### `optimizely.datafileRoute`
If you would like to see the latest configuration of your feature flags and experiments as defined in Optimizely's UI, you can see the datafile by installing this route on your server:
```
app.use('/optimizely/datafile', optimizely.datafileRoute);
```

## Demo
See the full [demo of usage](demo), reproduced below:
```
const optimizelyExpressSdk = require('@optimizely/express-sdk');
const optimizely = optimizelyExpressSdk.initialize({
  sdkKey: 'CZsVVgn6j9ce6fNPt2ZEiB',
  datafileOptions: {
    autoUpdate: true,
    updateInterval: 600000 // 10 minutes in milliseconds
  },
});

app.use(optimizely.middleware);

app.use('/webhooks/optimizely', bodyParser.text({ type: '*/*' }), optimizely.webhookRequest);

app.use('/optimizely/datafile', optimizely.datafileRoute);

app.get('/', function(req, res, next) {
  const isEnabled = req.optimizely.client.isFeatureEnabled('checkout_flow', '123');
  res.render('index', {
    title: 'Express: ' + (isEnabled ? 'feature on!' : 'feature off')
  });
});
```


## About

`@optimizely/express-sdk` is developed and maintained by [Optimizely](https://optimizely.com) and many [contributors](https://github.com/optimizely/express-sdk/graphs/contributors). If you're interested in learning more about what Optimizely X Full Stack can do for your company, please [get in touch](mailto:eng@optimizely.com)!


### Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md).

### Credits

This repository includes code from the following open source projects:

body-parser	<br/>
Copyright (c) 2014 Jonathan Ong <me@jongleberry.com> <br/>
Copyright (c) 2014-2015 Douglas Christopher Wilson <doug@somethingdoug.com> <br/>
License (MIT): https://github.com/expressjs/body-parser <br/>

cookie-parser <br/>
Copyright (c) 2014 TJ Holowaychuk <tj@vision-media.ca> <br/>
Copyright (c) 2015 Douglas Christopher Wilson <doug@somethingdoug.com> <br/>
License (MIT): https://github.com/expressjs/cookie-parser <br/>

debug <br/>
Copyright (c) 2014 TJ Holowaychuk <tj@vision-media.ca> <br/>
License (MIT): https://github.com/visionmedia/debug <br/>

express	<br/>
Copyright (c) 2009-2014 TJ Holowaychuk <tj@vision-media.ca> <br/>
Copyright (c) 2013-2014 Roman Shtylman <shtylman+expressjs@gmail.com> <br/>
Copyright (c) 2014-2015 Douglas Christopher Wilson <doug@somethingdoug.com> <br/>
License (MIT): https://github.com/expressjs/express <br/>

http-errors	<br/>
Copyright (c) 2014 Jonathan Ong me@jongleberry.com <br/>
Copyright (c) 2016 Douglas Christopher Wilson doug@somethingdoug.com <br/>
License (MIT): https://github.com/jshttp/http-errors <br/>

jade <br/>
Copyright (c) 2009-2014 TJ Holowaychuk <tj@vision-media.ca> <br/>
License (MIT): https://github.com/dscape/jade <br/>

morgan <br/>
Copyright (c) 2014 Jonathan Ong <me@jongleberry.com> <br/>
Copyright (c) 2014-2017 Douglas Christopher Wilson <doug@somethingdoug.com> <br/>
License (MIT): https://github.com/expressjs/morgan <br/>

### Additional Code

This software may be used with additional code that is separately downloaded by you. These components are subject to their own license terms, which you should review carefully.

request <br/>
Copyright (c)  2010-2012 Mikeal Rogers <br/>
License (Apache 2.0): https://github.com/request/request <br/>

request-promise <br/>
Copyright (c)  2017, Nicolai Kamenzky, Ty Abonil, and contributors <br/>
License (ISC): https://github.com/request/request-promise <br/>



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.0] - May 29, 2019

### New Features
- Initial release of the Optimizely Express SDK
- Features Include:
    - Middleware that adds Optimizely to every route
    - A secure webhook implementation
    - A route to view the latest datafile loaded by the middleware


index.js
/**
 * Optimizely Express SDK
 *
 * Copyright 2019, Optimizely
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict'

const OptimizelySdk = require('@optimizely/optimizely-sdk');
const { DatafileManager } = require('@optimizely/js-sdk-datafile-manager');
const crypto = require('crypto');
const rp = require('request-promise');


/**
 * optimizely
 *
 * Middelware which initializes and installs the Optimizely SDK onto an express request object
 *
 * @param {Object} options
 * @param {Object} options.sdkKey SDK key for your Optimizely project's environment
 * @param {Object} options.datafile initial datafile if you want to provide specific defaults on startup
 * @param {Object} options.logLevel log level for the default logger
 *
 * @returns {Function} to handle the express request
 */
function initialize(options) {

  let {
    sdkKey,
    datafile,
    logLevel,
  } = options;


  OptimizelySdk.setLogLevel(logLevel)
  OptimizelySdk.setLogger(OptimizelySdk.logging.createLogger())

  const manager = new DatafileManager({
    sdkKey,
    ...options
  });

  function updateDatafile() {
    datafile = manager.get()
    datafile._lastUpdated = new Date();
    console.log('[Optimizely] Datafile Updated!');
  }

  manager.on('update', updateDatafile);
  manager.onReady().then(updateDatafile);

  manager.start();

  return {
    /**
     * Optimizely Middleware
     * Provides an Optimizely client instance of the SDK available
     * on routes
     *
     * @param {Object} req express request object
     * @param {Object} res express response object
     * @param {Function} next express routing next function
     */
    middleware(req, res, next) {
      const optimizelyClient = OptimizelySdk.createInstance({
        datafile: datafile,
        ...options
      });

      req.optimizely = {
        datafile: datafile || {},
        client: optimizelyClient,
      }

      next();
    },

    /**
     * Optimizely Webhook Route
     * Route to accept webhook notifications from Optimizely
     *
     * @param {Object} req express request object
     * @param {Object} res express response object
     * @param {Function} next express routing next function
     */
    async webhookRequest(req, res, next) {

      const WEBHOOK_SECRET = process.env.OPTIMIZELY_WEBHOOK_SECRET
      const webhook_payload = req.body
      const hmac = crypto.createHmac('sha1', WEBHOOK_SECRET)
      const webhookDigest = hmac.update(webhook_payload).digest('hex')

      const computedSignature = `sha1=${webhookDigest}`
      const requestSignature = req.header('X-Hub-Signature')

      if (!crypto.timingSafeEqual(Buffer.from(computedSignature, 'utf8'), Buffer.from(requestSignature, 'utf8'))) {
        console.log(`[Optimizely] Signatures did not match! Do not trust webhook request")`)
        res.status(500)
        return
      }

      console.log(`
        [Optimizely] Optimizely webhook request received!
        Signatures match! Webhook verified as coming from Optimizely
        Download Optimizely datafile and re-instantiate the SDK Client
        For the latest changes to take affect
      `);

      const datafileString = await rp(`https://cdn.optimizely.com/datafiles/${sdkKey}.json`)
      datafile = JSON.parse(datafileString);

      res.sendStatus(200)
    },

    /**
     * datafileRoute
     *
     * Provides a route that exposes the contents of the datafile currently loaded in your application
     *
     * @param {Object} req express request object
     * @param {Object} res express response object
     * @param {Function} next express routing next function
     */
    datafileRoute(req, res, next) {
      const datafile = (req && req.optimizely && req.optimizely.datafile) || {}
      res.setHeader('Content-Type', 'application/json');
      res.status(200).send(JSON.stringify(datafile, null, '  '));
    },

    /**
     * isRouteEnabled (EXPERIMENTAL-FEATURE)
     *
     * Provides a method which can be used to block a route in express on whether the feature is enabled or not
     *
     * @param {String} featureKey for the specific feature in question
     * @param {Function} onRouteDisabled function called when the feature is disabled
     * @param {Error} featureKey for the specific feature in question
     *
     * @returns {Function}
     */
    isRouteEnabled(featureKey, onRouteDisabled) {
      return function (req, res, next) {
        // TODO: Improve design of user Id
        const userId = req.userId || 'test123'
        const optimizelyClient = req && req.optimizely && req.optimizely.client
        if (optimizelyClient) {
          // TODO: Pass in attributes
          const enabled = optimizelyClient.isFeatureEnabled(featureKey, userId);
          if (enabled) {
            // Feature is enabled move on to next route
            next();
            return
          }
        }
        onRouteDisabled(req, res, next);
      }
    }
  }
}


module.exports = {
  initialize,
}


demo/express-sdk.js
/**
 * Optimizely Express SDK
 *
 * Copyright 2019, Optimizely
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict'

const OptimizelySdk = require('@optimizely/optimizely-sdk');
const { DatafileManager } = require('@optimizely/js-sdk-datafile-manager');
const crypto = require('crypto');
const rp = require('request-promise');


/**
 * optimizely
 *
 * Middelware which initializes and installs the Optimizely SDK onto an express request object
 *
 * @param {Object} options
 * @param {Object} options.sdkKey SDK key for your Optimizely project's environment
 * @param {Object} options.datafile initial datafile if you want to provide specific defaults on startup
 * @param {Object} options.logLevel log level for the default logger
 *
 * @returns {Function} to handle the express request
 */
function initialize(options) {

  let {
    sdkKey,
    datafile,
    logLevel,
  } = options;


  OptimizelySdk.setLogLevel(logLevel)
  OptimizelySdk.setLogger(OptimizelySdk.logging.createLogger())

  const manager = new DatafileManager({
    sdkKey,
    ...options
  });

  function updateDatafile() {
    datafile = manager.get()
    datafile._lastUpdated = new Date();
    console.log('[Optimizely] Datafile Updated!');
  }

  manager.on('update', updateDatafile);
  //manager.onReady().then(updateDatafile);

  manager.start();

  return {
    /**
     * Optimizely Middleware
     * Provides an Optimizely client instance of the SDK available
     * on routes
     *
     * @param {Object} req express request object
     * @param {Object} res express response object
     * @param {Function} next express routing next function
     */
    middleware(req, res, next) {
      const optimizelyClient = OptimizelySdk.createInstance({
        datafile: datafile,
        ...options
      });

      req.optimizely = {
        datafile: datafile || {},
        client: optimizelyClient,
      }

      next();
    },

    /**
     * Optimizely Webhook Route
     * Route to accept webhook notifications from Optimizely
     *
     * @param {Object} req express request object
     * @param {Object} res express response object
     * @param {Function} next express routing next function
     */
    async webhookRequest(req, res, next) {

      const WEBHOOK_SECRET = process.env.OPTIMIZELY_WEBHOOK_SECRET
      const webhook_payload = req.body
      const hmac = crypto.createHmac('sha1', WEBHOOK_SECRET)
      const webhookDigest = hmac.update(webhook_payload).digest('hex')

      const computedSignature = `sha1=${webhookDigest}`
      const requestSignature = req.header('X-Hub-Signature')

      if (!crypto.timingSafeEqual(Buffer.from(computedSignature, 'utf8'), Buffer.from(requestSignature, 'utf8'))) {
        console.log(`[Optimizely] Signatures did not match! Do not trust webhook request")`)
        res.status(500)
        return
      }

      console.log(`
        [Optimizely] Optimizely webhook request received!
        Signatures match! Webhook verified as coming from Optimizely
        Download Optimizely datafile and re-instantiate the SDK Client
        For the latest changes to take affect
      `);

      const datafileString = await rp(`https://cdn.optimizely.com/datafiles/${sdkKey}.json`)
      datafile = JSON.parse(datafileString);

      res.sendStatus(200)
    },

    /**
     * datafileRoute
     *
     * Provides a route that exposes the contents of the datafile currently loaded in your application
     *
     * @param {Object} req express request object
     * @param {Object} res express response object
     * @param {Function} next express routing next function
     */
    datafileRoute(req, res, next) {
      const datafile = (req && req.optimizely && req.optimizely.datafile) || {}
      res.setHeader('Content-Type', 'application/json');
      res.status(200).send(JSON.stringify(datafile, null, '  '));
    },

    /**
     * isRouteEnabled
     *
     * Provides a method which can be used to block a route in express on whether the feature is enabled or not
     *
     * @param {String} featureKey for the specific feature in question
     * @param {Function} onRouteDisabled function called when the feature is disabled
     * @param {Error} featureKey for the specific feature in question
     *
     * @returns {Function}
     */
    isRouteEnabled(featureKey, onRouteDisabled) {
      return function (req, res, next) {
        // TODO: Improve design of user Id
        const userId = req.userId || 'test123'
        const optimizelyClient = req && req.optimizely && req.optimizely.client
        if (optimizelyClient) {
          // TODO: Pass in attributes
          const enabled = optimizelyClient.isFeatureEnabled(featureKey, userId);
          if (enabled) {
            // Feature is enabled move on to next route
            next();
            return
          }
        }
        onRouteDisabled(req, res, next);
      }
    }
  }
}


module.exports = {
  initialize,
}


demo/app.js
const createError = require('http-errors');
const express = require('express');
const path = require('path');
const cookieParser = require('cookie-parser');
const logger = require('morgan');
const bodyParser = require('body-parser');

const router = express.Router();
const app = express();

// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'jade');

/*************
 * Optimizely Express SDK Usage
 */
// TODO: Update once package is public
//const optimizelyExpressSdk = require('../index.js');
const optimizelyExpressSdk = require('./express-sdk.js');
const optimizely = optimizelyExpressSdk.initialize({
  sdkKey: 'CZsVVgn6j9ce6fNPt2ZEiB',
  datafileOptions: {
    autoUpdate: true,
    updateInterval: 10*1000 // 10 seconds in milliseconds
  },
});
app.use(optimizely.middleware);
app.use('/webhooks/optimizely', bodyParser.text({ type: '*/*' }), optimizely.webhookRequest);
app.use('/optimizely/datafile', optimizely.datafileRoute);

app.get('/', function(req, res, next) {
  const isEnabled = req.optimizely.client.isFeatureEnabled('checkout_flow', '123');
  res.render('index', {
    title: 'Express: ' + (isEnabled ? 'feature on!' : 'feature off')
  });
});

/*************/

app.use(logger('dev'));
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));

// catch 404 and forward to error handler
app.use(function(req, res, next) {
  next(createError(404));
});

// error handler
app.use(function(err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.render('error');
});

module.exports = app;
